shell/platform/windows/compositor_opengl.cc

```diff
@@ -0,0 +1,147 @@
+  return true;
+}
+
+bool CompositorOpenGL::CollectBackingStore(const FlutterBackingStore* store) {
+  FML_DCHECK(is_initialized_);
+  FML_DCHECK(store->type == kFlutterBackingStoreTypeOpenGL);
+  FML_DCHECK(store->open_gl.type == kFlutterOpenGLTargetTypeFramebuffer);
+
+  auto user_data = static_cast<FramebufferBackingStore*>(
+      store->open_gl.framebuffer.user_data);
+
+  gl_->DeleteFramebuffers(1, &user_data->framebuffer_id);
+  gl_->DeleteTextures(1, &user_data->texture_id);
+
+  delete user_data;
+  return true;
+}
+
+bool CompositorOpenGL::Present(const FlutterLayer** layers,
+                               size_t layers_count) {
+  FML_DCHECK(is_initialized_);
+  FML_DCHECK(layers_count == 1);
+  FML_DCHECK(layers[0]->type == kFlutterLayerContentTypeBackingStore);
+  FML_DCHECK(layers[0]->backing_store->type == kFlutterBackingStoreTypeOpenGL);
+  FML_DCHECK(layers[0]->backing_store->open_gl.type ==
+             kFlutterOpenGLTargetTypeFramebuffer);
+
+  if (!engine_->view()) {
+    return false;
+  }
+
+  auto width = layers[0]->size.width;
+  auto height = layers[0]->size.height;
+
+  // Acquiring the view's framebuffer ID resizes its surface if necessary.
+  auto destination_id = engine_->view()->GetFrameBufferId(width, height);
+  auto source_id = layers[0]->backing_store->open_gl.framebuffer.name;
+
+  if (!engine_->surface_manager()->MakeCurrent()) {
+    return false;
+  }
+
+  gl_->BindFramebuffer(GL_READ_FRAMEBUFFER, source_id);
+  gl_->BindFramebuffer(GL_DRAW_FRAMEBUFFER, destination_id);
+
+  gl_->BlitFramebuffer(0,                    // srcX0
+                       0,                    // srcY0
+                       width,                // srcX1
+                       height,               // srcY1
+                       0,                    // dstX0
+                       0,                    // dstY0
+                       width,                // dstX1
+                       height,               // dstY1
+                       GL_COLOR_BUFFER_BIT,  // mask
+                       GL_NEAREST            // filter
+  );
+
+  return engine_->view()->SwapBuffers();
+}
+
+bool CompositorOpenGL::Initialize() {
+  FML_DCHECK(!is_initialized_);
+
+  if (!engine_->surface_manager()->MakeCurrent()) {
+    return false;
+  }
+
+  auto gl = std::make_unique<impeller::ProcTableGLES>(resolver_);
+  if (!gl->IsValid()) {
+    return false;
+  }
+
+  // Based off Skia's logic:
+  // https://github.com/google/skia/blob/4738ed711e03212aceec3cd502a4adb545f38e63/src/gpu/ganesh/gl/GrGLCaps.cpp#L1963-L2116
+  auto description = gl->GetDescription();
+  if (description->HasExtension("GL_EXT_texture_format_BGRA8888")) {
+    format_ = GL_BGRA8_EXT;
+  } else if (description->HasExtension("GL_APPLE_texture_format_BGRA8888") &&
+             description->GetGlVersion().IsAtLeast(impeller::Version(3, 0))) {
+    format_ = GL_BGRA8_EXT;
+  } else {
+    format_ = GL_RGBA8;
+  }
+
+  gl_ = std::move(gl);
```

Comment:
This is totally fine; you could also just initialise the field directly and `.reset()` it if `!gl_->IsValid()`. That has the very minor advantage of keeping the initialisation and error handling for failed initialisation all in one place and avoids an extra local/move, but I don't feel at all strongly about it.

Reply:
Updated!